esp-hal-common/src/interrupt/riscv.rs

```diff
@@ -536,92 +494,87 @@ unsafe fn handle_exception(pc: usize, trap_frame: *mut TrapFrame) {
-    (*trap_frame).a6 = atomic_emulation_trap_frame.a6;
-    (*trap_frame).a7 = atomic_emulation_trap_frame.a7;
-    (*trap_frame).s2 = atomic_emulation_trap_frame.s2;
-    (*trap_frame).s3 = atomic_emulation_trap_frame.s3;
-    (*trap_frame).s4 = atomic_emulation_trap_frame.s4;
-    (*trap_frame).s5 = atomic_emulation_trap_frame.s5;
-    (*trap_frame).s6 = atomic_emulation_trap_frame.s6;
-    (*trap_frame).s7 = atomic_emulation_trap_frame.s7;
-    (*trap_frame).s8 = atomic_emulation_trap_frame.s8;
-    (*trap_frame).s9 = atomic_emulation_trap_frame.s9;
-    (*trap_frame).s10 = atomic_emulation_trap_frame.s10;
-    (*trap_frame).s11 = atomic_emulation_trap_frame.s11;
-    (*trap_frame).t3 = atomic_emulation_trap_frame.t3;
-    (*trap_frame).t4 = atomic_emulation_trap_frame.t4;
-    (*trap_frame).t5 = atomic_emulation_trap_frame.t5;
-    (*trap_frame).t6 = atomic_emulation_trap_frame.t6;
+    let mut frame = [
+        (*trap_frame).ra,
+        (*trap_frame).sp,
+        (*trap_frame).gp,
+        (*trap_frame).tp,
+        (*trap_frame).t0,
+        (*trap_frame).t1,
+        (*trap_frame).t2,
+        (*trap_frame).s0,
+        (*trap_frame).s1,
+        (*trap_frame).a0,
+        (*trap_frame).a1,
+        (*trap_frame).a2,
+        (*trap_frame).a3,
+        (*trap_frame).a4,
+        (*trap_frame).a5,
+        (*trap_frame).a6,
+        (*trap_frame).a7,
+        (*trap_frame).s2,
+        (*trap_frame).s3,
+        (*trap_frame).s4,
+        (*trap_frame).s5,
+        (*trap_frame).s6,
+        (*trap_frame).s7,
+        (*trap_frame).s8,
+        (*trap_frame).s9,
+        (*trap_frame).s10,
+        (*trap_frame).s11,
+        (*trap_frame).t3,
+        (*trap_frame).t4,
+        (*trap_frame).t5,
+        (*trap_frame).t6,
+        (*trap_frame).pc,
+    ];
+
+    riscv_atomic_emulation_trap::atomic_emulation((*trap_frame).pc, &mut frame);
+
+    (*trap_frame).ra = frame[0];
+    (*trap_frame).sp = frame[1];
+    (*trap_frame).gp = frame[2];
+    (*trap_frame).tp = frame[3];
+    (*trap_frame).t0 = frame[4];
+    (*trap_frame).t1 = frame[5];
+    (*trap_frame).t2 = frame[6];
+    (*trap_frame).s0 = frame[7];
+    (*trap_frame).s1 = frame[8];
+    (*trap_frame).a0 = frame[9];
+    (*trap_frame).a1 = frame[10];
+    (*trap_frame).a2 = frame[11];
+    (*trap_frame).a3 = frame[12];
+    (*trap_frame).a4 = frame[13];
+    (*trap_frame).a5 = frame[14];
+    (*trap_frame).a6 = frame[15];
+    (*trap_frame).a7 = frame[16];
+    (*trap_frame).s2 = frame[17];
+    (*trap_frame).s3 = frame[18];
+    (*trap_frame).s4 = frame[19];
+    (*trap_frame).s5 = frame[20];
+    (*trap_frame).s6 = frame[21];
+    (*trap_frame).s7 = frame[22];
+    (*trap_frame).s8 = frame[23];
+    (*trap_frame).s9 = frame[24];
+    (*trap_frame).s10 = frame[25];
+    (*trap_frame).s11 = frame[26];
+    (*trap_frame).t3 = frame[27];
+    (*trap_frame).t4 = frame[28];
+    (*trap_frame).t5 = frame[29];
+    (*trap_frame).t6 = frame[30];
+    (*trap_frame).pc = frame[31];
```

Comment:
I think we need to advance the PC by 4 bytes since `riscv_atomic_emulation_trap::atomic_emulation` won't do that?